csharp/src/Ice/Communicator.cs

```diff
@@ -850,46 +849,35 @@ public async ValueTask<IObjectPrx> CreateAdminAsync(ObjectAdapter? adminAdapter,
                     throw new InvalidOperationException("Admin is disabled");
                 }
 
-                _adminIdentity = adminIdentity;
                 if (adminAdapter == null)
                 {
-                    if (GetProperty("Ice.Admin.Endpoints") != null)
-                    {
-                        adminAdapter = CreateObjectAdapter("Ice.Admin");
-                    }
-                    else
+                    if (GetProperty("Ice.Admin.Endpoints") == null)
                     {
-                        throw new InvalidConfigurationException("Ice.Admin.Endpoints is not set");
+                        throw new InvalidConfigurationException("Ice.Admin has no endpoint");
                     }
+
+                    _adminIdentity = adminIdentity;
+                    _createAdminTask = CreateAdminAsync(cancel);
+                    createAdminTask = _createAdminTask;
                 }
                 else
                 {
+                    _adminIdentity = adminIdentity;
                     _adminAdapter = adminAdapter;
+                    AddAllAdminFacets();
                 }
-                Debug.Assert(_adminAdapter != null);
-                AddAllAdminFacets();
             }
 
-            if (adminAdapter == null) // the parameter is null which means _adminAdapter needs to be activated
+            if (createAdminTask != null)
             {
-                try
-                {
-                    await _adminAdapter.ActivateAsync().ConfigureAwait(false);
-                }
-                catch
-                {
-                    // We cleanup _adminAdapter, however this error is not recoverable (can't call again GetAdmin()
-                    // after fixing the problem) since all the facets (servants) in the adapter are lost
-                    await _adminAdapter.DisposeAsync().ConfigureAwait(false);
-                    lock (_mutex)
-                    {
-                        _adminAdapter = null;
-                    }
-                    throw;
-                }
+                return await createAdminTask.ConfigureAwait(false);
```

Comment:
Should we cleanup the `_createAdminTask` in case of a failure, so that you can retry?

Reply:
No, because all failures correspond to bug, and applications should not expect bugs and then try again with non-buggy code/configuration.